### PR TITLE
Add Simple Factory Pattern implementation

### PR DESCRIPTION
**Pull request by samilAyoub**
Java source code demonstrate simple factory design pattern.

relates to : #1477

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -195,6 +195,7 @@
         <module>arrange-act-assert</module>
         <module>transaction-script</module>
         <module>filterer</module>
+        <module>simple-factory</module>
```
Comment (iluwatar) [inline pom.xml:198]: Please use the name `factory`

File: simple-factory/src/main/java/com/iluwatar/simplefactory/App.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.iluwatar.simplefactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
Comment (iluwatar) [inline simple-factory/src/main/java/com/iluwatar/simplefactory/App.java:28]: Please describe here the pattern in few sentences and how this example implements it. See the other patterns for examples.

File: simple-factory/src/main/java/com/iluwatar/simplefactory/App.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+public class App {
+  
+  private static final Logger LOGGER = LoggerFactory.getLogger(App.class);
+  
+  /**
+   * Program main entry point.
+   */
+  public static void main(String[] args) {
+    Car car1 = CarSimpleFactory.getCar(CarSimpleFactory.CarType.FORD);
```
Comment (iluwatar) [inline simple-factory/src/main/java/com/iluwatar/simplefactory/App.java:37]: Please use local variable type inference `var` where possible

**Review by iluwatar — CHANGES_REQUESTED** [review]
`README.md` is missing

**Comment by samilAyoub** [comment]
@iluwatar I dont know why urm-maven-plugin generates factory.urm.puml but does not genertaes factory.urm.png?

File: pom.xml (hunk 1 above)
Comment (samilAyoub) [reply pom.xml:198]: Done

**Comment by ravening** [comment]
One suggestion which I can say is to use Lambda expressions instead of using switch cases to generate Cars.
You can refer to an example in https://github.com/ravening/Lambda-Expressions_java/tree/master/designpatterns/rv/factory

File: factory/pom.xml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+    </dependency>
+  </dependencies>
+  <build>
+		<plugins>
+			<!-- Maven assembly plugin is invoked with default setting which we have
```
Comment (ravening) [inline factory/pom.xml:22]: Better to use spaces instead of tabs

Comment (samilAyoub) [reply factory/pom.xml]: Done

Comment (samilAyoub) [reply factory/pom.xml]: @ravening Done !

Comment (samilAyoub) [reply factory/pom.xml]: @ravening please review the changes.

**Comment by iluwatar** [comment]
> @iluwatar I don't know why urm-maven-plugin generates factory.urm.puml but does not generates factory.urm.png?

It doesn't, the image needs to be created by hand. See https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute

File: factory/README.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+> Lets say we have a web application connected to SQLServer, but now we want to switch to Oracle. To do so without modifying existing source code, we need to implements Simple Factory pattern, in which a static method can be invoked to create connection to a given database.
+
+Wikipedia says
+
+> Factory is an object for creating other objects – formally a factory is a function or method that returns objects of a varying prototype or class.
+
+**Programmatic Example**
+
+We have an interface "Car" and tow implementations "Ford" and "Ferrari".
```
Comment (iluwatar) [inline factory/README.md:32]:
```suggestion
We have an interface `Car` and two implementations `Ford` and `Ferrari`.
```

File: factory/README.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  CarType(Supplier<Car> constructor) {
+    this.constructor = constructor;
+  }
+  
+  public Supplier<Car> getConstructor() {
+    return this.constructor;
+  }
+}
+```
+Then we have the static method "getCar" to create car objects encapsulated in the factory class "CarSimpleFactory".
```
Comment (iluwatar) [inline factory/README.md:93]:
```suggestion
Then we have the static method `getCar` to create car objects encapsulated in the factory class `CarSimpleFactory`.
```

File: factory/README.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+This is Ford.
+This Ferrari.
+```
+## Applicability
+
+Use the Simple Factory pattern when you only care about the creation of a object, not how to create and manage it.
+
+## Disadvantages:
+
+The code becomes more complicated than it should be.
```
Comment (iluwatar) [inline factory/README.md:131]: We could rephrase this part a bit so that there is ##Applicability section and under it `Pros` and `Cons`.

Comment (iluwatar) [reply factory/README.md:131]: Additionally please add a class diagram.

**Review by iluwatar — CHANGES_REQUESTED** [review]

File: factory/README.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+> Lets say we have a web application connected to SQLServer, but now we want to switch to Oracle. To do so without modifying existing source code, we need to implements Simple Factory pattern, in which a static method can be invoked to create connection to a given database.
+
+Wikipedia says
+
+> Factory is an object for creating other objects – formally a factory is a function or method that returns objects of a varying prototype or class.
+
+**Programmatic Example**
+
+We have an interface 'Car' and tow implementations 'Ford' and 'Ferrari'.
```
Comment (iluwatar) [inline factory/README.md:32]:
```suggestion
We have an interface `Car` and tow implementations `Ford` and `Ferrari`.
```

File: factory/README.md (hunk 5 above)
Comment (iluwatar) [reply factory/README.md]: Still not ok

File: factory/README.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+This is Ford.
+This Ferrari.
+```
+## Class Diagram
+![alt text](./etc/factory.urm.png "Factory pattern class diagram")
+
+## Applicability
+Use the Simple Factory pattern when you only care about the creation of a object, not how to create and manage it.
+
+## Pros
```
Comment (iluwatar) [inline factory/README.md:131]:
```suggestion
Pros
```

File: factory/README.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+* Allows keeping all objects creation in one place and avoid of spreading 'new' key value across codebase.
+* Allows to writs loosely coupled code. Some of its main advantages include better testability, easy-to-understand code, swappable components, scalability and isolated features.
+
+## Cons
```
Comment (iluwatar) [inline factory/README.md:135]:
```suggestion
Cons
```

**Review by iluwatar — APPROVED** [review]

**Comment by iluwatar** [comment]
Found just minor issues so I'll merge it and push a change after it.

**Comment by iluwatar** [comment]
Many thanks for the new pattern! 

@all-contributors please add @samilAyoub for code

**Comment by iluwatar** [comment]
@all-contributors please add @ravening for review

**Pull request merged**